flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-doris/src/main/java/com/ververica/cdc/connectors/doris/sink/DorisEventSerializer.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2023 Ververica Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ververica.cdc.connectors.doris.sink;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.types.DataType;
+
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.core.JsonProcessingException;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectMapper;
+
+import com.ververica.cdc.common.data.GenericRecordData;
+import com.ververica.cdc.common.data.RecordData;
+import com.ververica.cdc.common.event.CreateTableEvent;
+import com.ververica.cdc.common.event.DataChangeEvent;
+import com.ververica.cdc.common.event.Event;
+import com.ververica.cdc.common.event.OperationType;
+import com.ververica.cdc.common.event.SchemaChangeEvent;
+import com.ververica.cdc.common.event.TableId;
+import com.ververica.cdc.common.schema.Column;
+import com.ververica.cdc.common.schema.Schema;
+import com.ververica.cdc.common.types.utils.DataTypeUtils;
+import com.ververica.cdc.common.utils.Preconditions;
+import com.ververica.cdc.common.utils.RecordDataUtils;
+import com.ververica.cdc.common.utils.SchemaUtils;
+import org.apache.doris.flink.deserialization.converter.DorisRowConverter;
+import org.apache.doris.flink.sink.writer.serializer.DorisRecordSerializer;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.doris.flink.sink.util.DeleteOperation.addDeleteSign;
+
+/** A serializer for Event to Tuple2<String, byte[]> */
+public class DorisEventSerializer implements DorisRecordSerializer<Event> {
+    private ObjectMapper objectMapper = new ObjectMapper();
+    private Map<TableId, Schema> schemaMaps = new HashMap<>();
```

Comment:
A troublesome problem is that we need to maintain the schemaMaps in State to recover from failure, so we need to  add a subclass of [DorisWriter](https://github.com/apache/doris-flink-connector/blob/1e2965d17d6e2478183a5dff955a6895c1ec9678/flink-doris-connector/src/main/java/org/apache/doris/flink/sink/writer/DorisWriter.java#L218) or [DorisBatchWriter](https://github.com/apache/doris-flink-connector/blob/1e2965d17d6e2478183a5dff955a6895c1ec9678/flink-doris-connector/src/main/java/org/apache/doris/flink/sink/batch/DorisBatchWriter.java#L39) to overwrite `snapshotState` method.
What do you think?

Reply:
Sorry for this comment, as the discuss before, CreateTableEvent will always sent before DataChangEvent, so we don't need to consider this situation.

Reply:
Yes, I understand that it will resend CreateTableEvent even if it is restarted.